### ISSUE TITLE
Loss API: unify forward+backward reduction semantics for microbatching with batch dimension

### ISSUE THREAD

**Issue by LeoBuron** [es-ude/OnDeviceTraining#135]
## Context

The forward and backward halves of every loss function have **inconsistent reduction semantics**, and the asymmetry will harden into a real bug when ODT moves microbatching from \`B = 1\` to \`B ≥ 1\` (i.e., adds a batch dimension to the tensors that flow through layers).

Today (post-#90):

| Stage | Signature | Reduction-aware? | Behaviour |
|---|---|---|---|
| \`mseLossForward(output, label)\` | 2 args | no | always \`sum_squared / numElements\` (mean over elements) |
| \`crossEntropyForwardFloat(output, label)\` | 2 args | no | always \`Σ y · -log p\` (no reduction) |
| \`mseLossBackward(output, label, result, batchSize, reduction)\` | 5 args | yes | divisor = \`batchSize\` for MEAN, 1 for SUM |
| \`crossEntropySoftmaxBackward(output, dist, loss, batchSize, reduction)\` | 5 args | yes | same |
| \`TrainingBatchDefault.c:20\` aggregator | hardcoded | **no** | \`return totalLoss / batch->size\` regardless of reduction |
| \`evaluationBatch\` aggregator | hardcoded | n/a (always mean) | same hardcoded division |

Two real consequences today:

1. **\`reduction = SUM\` is silently broken** at the caller layer: \`TrainingBatchDefault\` divides by \`batch->size\` even for SUM, producing a per-sample mean instead of a sum. Latent because no test currently uses SUM.
2. **Forward and backward speak different languages**. Forward returns raw per-sample loss; backward applies macro-scaling on top. Reading \`trainingStats->loss\` directly gives a per-sample number; reading \`mseLossBackward(...)\`'s gradient gives a macro-scaled gradient.

Both are tolerable while microbatching is \`B = 1\` and \`reduction\` is always MEAN. They become incoherent the moment a microbatch holds more than one sample, because then \"is this scaled by the microbatch size, the macro batch size, or neither?\" must have a single answer.

## Three batching levels (terminology)

| Level | Tensor shape | Where it lives | Notes |
|---|---|---|---|
| Sample | \`[F]\` (or never materialised separately) | conceptual only | one example |
| Microbatch | \`[B, F]\` with \`B ≥ 1\` | layer forward/backward, loss forward/backward | what one model invocation processes; \`B = 1\` today, \`B ≥ 1\` future |
| Macro batch | implicit \`[N, ...]\` realised as \`K\` microbatches | \`TrainingBatchDefault\`, \`evaluationBatch\` | \`N = Σ B_k\`; gradient accumulation across microbatches |

The current code uses \"batchSize\" to mean the **macro** count when passed to backward, and \"microbatch\" when looking at tensor shapes (always \`B = 1\` today). This issue proposes settling that ambiguity.

### Forward and backward become symmetric

Both halves of every loss function take \`(macroBatchSize, reduction)\` and apply the macro-scaling at micro-level — exactly mirroring what backward already does.

\`\`\`c
/* Forward — reduction applied at microbatch level, scaled to macro semantics. */
float mseLossForward(tensor_t *output, tensor_t *label,
                     size_t macroBatchSize, reduction_t reduction);

float crossEntropyForward(tensor_t *softmaxOutput, tensor_t *distribution,
                          size_t macroBatchSize, reduction_t reduction);

/* Backward — semantics unchanged; parameter renamed for clarity. */
void mseLossBackward(tensor_t *modelOutput, tensor_t *label, tensor_t *result,
                     size_t macroBatchSize, reduction_t reduction);

void crossEntropySoftmaxBackward(tensor_t *softmaxOutput, tensor_t *distribution,
                                 tensor_t *loss,
                                 size_t macroBatchSize, reduction_t reduction);
\`\`\`

Forward sketch (MSE):

\`\`\`c
float mseLossForward(output, label, macroBatchSize, reduction) {
    /* per-sample squared error summed across the microbatch */
    float perMicroSum = 0.0f;
    for each sample b in microbatch (output->shape->dimensions[0]):
        for each feature f:
            perMicroSum += (output[b,f] - label[b,f]) * (output[b,f] - label[b,f]);
    /* mean-over-features convention (see open question 1 below) */
    perMicroSum /= numFeaturesPerSample;

    return (reduction == REDUCTION_MEAN)
        ? perMicroSum / (float)macroBatchSize
        : perMicroSum;
}
\`\`\`

Backward unchanged behaviour, just the explicit naming.

### Aggregator becomes trivial

\`TrainingBatchDefault\` no longer applies any reduction division — forward already pre-scales:

\`\`\`c
float trainingBatchDefault(model, modelSize, lossConfig, batch, calculateGradsFn) {
    float totalLoss = 0.0f;
    for each microbatch mb in batch:
        stats = calculateGradsFn(model, modelSize, lossConfig,
                                 batch->totalSamples,  /* macroBatchSize */
                                 mb->item, mb->label);
        totalLoss += stats->loss;
    return totalLoss;
}
\`\`\`

\`evaluationBatch\` similar, hard-wired to MEAN reduction (or made reduction-configurable as a follow-up).

### Symmetry properties this delivers

- **Forward and backward both take the same scaling args.** No more \"backward knows the batch but forward doesn't\".
- **Per-microbatch loss already reflects macro reduction.** Caller just sums.
- **Gradient accumulation works naturally.** With \`K\` microbatches of size \`B_k\`, summing \`K\` per-microbatch losses gives \`(1/N) Σ_all L_kb\` for MEAN or \`Σ_all L_kb\` for SUM, automatically.
- **Migration path is transparent.** \`B = 1\` today still works; future \`B ≥ 1\` is a tensor-shape change, not an API change.

## Breaking changes (accepted)

| File | Change |
|---|---|
| \`src/loss_functions/include/MSE.h\`, \`CrossEntropy.h\`, \`LossFunction.h\` | New forward signatures; \`lossFwdFn_t\` typedef updated |
| \`src/loss_functions/MSE.c\`, \`CrossEntropy.c\` | Forward implementations apply reduction |
| \`src/loss_functions/LossFunction.c\` | \`lossFunctions[]\` table entries match new signatures |
| \`src/userApi/InferenceApi.c\` | \`inferenceWithLoss\` accepts \`(macroBatchSize, reduction)\` and threads them through |
| \`src/userApi/training_loop/calculate_grads/CalculateGradsSequential.c\` | Forward call updated to pass macro args |
| \`src/userApi/training_loop/training_batch/TrainingBatchDefault.c\` | Drop hardcoded \`/ batch->size\`; aggregation becomes plain sum |
| \`src/userApi/training_loop/TrainingLoopApi.c::evaluationBatch\` | Same; pin reduction to MEAN explicitly |
| All loss tests under \`test/unit/loss_functions/\` | Update calls to new signatures |
| Tests that exercise \`inferenceWithLoss\`, \`trainingRun\`, \`evaluationBatch\` | Same |

## Open questions to resolve before implementation

1. **MSE-mean convention.** Should MSE-forward's mean be over all elements (samples × features, PyTorch \`reduction='mean'\`) or only over features per sample (current ad-hoc)? Recommendation: PyTorch convention (\`/numElements\`); document explicitly in \`LossFunction.h\` analogous to #95.

2. **Microbatch-shape contract.** Today loss forward sees \`[1, F]\`; future it sees \`[B, F]\`. Should the contract be documented in \`LossFunction.h\` (\`output->shape->dimensions[0]\` is the microbatch size, the loss iterates internally)? Recommendation: yes, single-sentence Doxygen note on \`lossFwdFn_t\` and \`lossBwdFn_t\` typedefs.

3. **\`evaluationBatch\` reduction parameter.** Hardcode MEAN (current behaviour, simplest), or expose to caller (consistent with training)? Recommendation: hardcode MEAN for v1 to keep the breaking-change scope tight; add a parameter as a follow-up if a real use case emerges.

4. **Per-sample vs. per-microbatch loss reporting.** With pre-scaled forward, \`trainingStats->loss\` is now a fractional macro contribution, not a readable per-sample value. Is that acceptable for diagnostics? Recommendation: yes — the readable per-sample value is recoverable as \`stats->loss * macroBatchSize\` if needed; the symmetry pay-off is worth the change.

## Related

- Closes the SUM-side latent bug noted in the comment thread on this issue (TrainingBatchDefault hardcoded \`/ batch->size\`); a quick-fix-only issue may be filed separately if the refactor lands later than the next training run that needs SUM.
- Audit pass 2 candidate (per the closing comment of #98 — \"batch-forward support in inference\" was queued for that pass).
- Memory entry \`Backward batchSize=1\` becomes stale once microbatching with batch dim lands; needs updating in tandem.

## Acceptance criteria

- [ ] Both halves of every loss function take \`(macroBatchSize, reduction)\` with the documented semantics.
- [ ] \`TrainingBatchDefault\` and \`evaluationBatch\` no longer apply any hardcoded division.
- [ ] A new test exercises \`reduction = REDUCTION_SUM\` end-to-end and verifies the returned macro loss is the literal sum of per-sample losses.
- [ ] An existing or new test exercises \`reduction = REDUCTION_MEAN\` end-to-end and verifies the returned macro loss is the literal mean of per-sample losses (numerically equivalent to current behaviour for \`MEAN\`).
- [ ] All existing tests under \`test/unit/loss_functions/\`, \`test/unit/userAPI/\` pass.
- [ ] Header doc on \`lossFwdFn_t\` / \`lossBwdFn_t\` documents the \`(macroBatchSize, reduction)\` contract and the microbatch-shape contract from open question 2.
